System/Win32/NLS.hsc

```diff
@@ -497,6 +499,54 @@ foreign import WINDOWS_CCONV unsafe "windows.h IsValidLocaleName"
 foreign import WINDOWS_CCONV unsafe "windows.h IsValidLocale"
   isValidLocale :: LCID -> LocaleTestFlags -> IO Bool
 
+type EnumLocalesFlag = DWORD
+
+-- The following locale enumeration flag constants are excluded from the `enum`
+-- list below, for the reason indicated:
+-- LOCALE_NEUTRALDATA -- Introduced in Windows 7 but not supported.
+
+#{enum EnumLocalesFlag,
+ , lOCALE_ALL             = LOCALE_ALL
+ , lOCALE_ALTERNATE_SORTS = LOCALE_ALTERNATE_SORTS
+ , lOCALE_REPLACEMENT     = LOCALE_REPLACEMENT
+ , lOCALE_SUPPLEMENTAL    = LOCALE_SUPPLEMENTAL
+ , lOCALE_WINDOWS         = LOCALE_WINDOWS
+ }
+
+type LOCALE_ENUMPROCEX = LPWSTR -> EnumLocalesFlag -> LPARAM -> IO BOOL
+foreign import WINDOWS_CCONV "wrapper"
+  mkLOCALE_ENUMPROCEX :: LOCALE_ENUMPROCEX -> IO (FunPtr LOCALE_ENUMPROCEX)
+
+enumSystemLocalesEx :: LOCALE_ENUMPROCEX -> EnumLocalesFlag -> LPARAM -> IO ()
+enumSystemLocalesEx callback dwFlags lParam = do
+  c_callback <- mkLOCALE_ENUMPROCEX callback
+  failIfFalse_ "EnumSystemLocalesEx" $
+    c_EnumSystemLocalesEx c_callback dwFlags lParam nullPtr
+  freeHaskellFunPtr c_callback
+foreign import WINDOWS_CCONV safe "windows.h EnumSystemLocalesEx"
+  c_EnumSystemLocalesEx :: (FunPtr LOCALE_ENUMPROCEX)
+                        -> DWORD
+                        -> LPARAM
+                        -> LPVOID
+                        -> IO Bool
+
+enumSystemLocalesEx' :: EnumLocalesFlag
+                     -> Maybe Bool
+                     -- ^ Maybe include (or exclude) replacement locales?
+                     -> IO [String]
+enumSystemLocalesEx' dwFlags mIsReplacement = do
+  store <- newIORef []
+  let localeEnumProcEx c_locale arg2 _ = do
+        locale <- peekTString c_locale
+        case mIsReplacement of
+          Nothing -> modifyIORef store (locale:)
+          Just isReplacement ->
+            when (isReplacement == (arg2 .&. lOCALE_REPLACEMENT /= 0)) $
+              modifyIORef store (locale:)
+        return True
+  enumSystemLocalesEx localeEnumProcEx dwFlags 0
+  reverse <$> readIORef store
```

Comment:
curious about the reverse, also why the `IORef`? can't this be done just by returning a list in each recursive call?

Reply:
The [callback function](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nc-winnls-locale_enumprocex) used by the Win32 API returns only `BOOL`, so it seemed to me that the only way (in Haskell) to get information out of the callback was via `IO BOOL`. I also found this StackOverflow QA useful: [What is a functional way to accumulate values from callback into a list?](https://stackoverflow.com/questions/23657450/what-is-a-functional-way-to-accumulate-values-from-callback-into-a-list)

As `(locale:)` (efficiently) adds each locale to the head of the building list, I used `reverse` to restore the final list to the same order as returned by the Win32 API (which is alphabetical).

I did wonder if I should use the [strict version](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-IORef.html#v:modifyIORef-39-) of `modifyIORef`, but it seemed unlikely a space leak would occur.

Reply:
Ah thanks! Had missed that it's a callback.